ssi-dids/src/did.rs

```diff
@@ -0,0 +1,392 @@
+use core::fmt;
+use std::ops::Deref;
+
+mod url;
+
+use serde::{Deserialize, Serialize};
+pub use url::*;
+
+/// Error raised when a conversion to a DID fails.
+#[derive(Debug, thiserror::Error)]
+#[error("invalid DID `{0}`: {1}")]
+pub struct InvalidDID<T>(pub T, pub Unexpected);
+
+impl<T> InvalidDID<T> {
+    pub fn map<U>(self, f: impl FnOnce(T) -> U) -> InvalidDID<U> {
+        InvalidDID(f(self.0), self.1)
+    }
+}
+
+/// DID.
+///
+/// This type is unsized and used to represent borrowed DIDs. Use `DIDBuf` for
+/// owned DIDs.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct DID([u8]);
```

Comment:
What is `bstr`? Or do you mean `str`? I'm used to using `[u8]` for pure ASCII strings because it simplifies many things (like indexing the string), but yes, I guess we can use `str` instead since it's a superset of ASCII.

Reply:
I meant https://crates.io/crates/bstr

Reply:
We don't care about UTF-8 here. DIDs only use ASCII characters, which are encoded the same regardless of the unicode encoding. I could use `str`, but `[u8]` is simpler to interact with. Maybe we could use https://docs.rs/ascii/1.1.0/ascii/struct.AsciiStr.html to make it clearer but I'm not sure its really necessary.

Reply:
Oh I just realized it may not be ASCII or UTF-8 at all! There is an option in the header to disable base64 encoding so the payload can be raw bytes.

Reply:
We could do that, but a typical design pattern in rust is to use DSTs for this kind of "specialized" strings. Just like there is `str`/`String`, `Path`\`PathBuf`, `OsStr`\`OsString`, `BStr`\`BString` in the `bstr` crate you showed me and `AsciiStr`\`AsciiString` in the `ascii` crate. It avoids a lot of unnecessary memory allocations. Here I did the same with `DID`\`DIDBuf`.

The comment "owned DIDs" above the `DID` type is wrong, I removed it. It is not owned but borrowed.